datanode/snapshot/service_create_snapshot.go

```diff
@@ -0,0 +1,269 @@
+		m.ChainID, m.HeightFrom, m.HeightTo, m.CurrentStateSnapshotFile, m.HistorySnapshotFile)
+}
+
+func (b *Service) CreateSnapshotAsync(ctx context.Context, chainID string, fromHeight int64, toHeight int64) (Meta, error) {
+	return b.createSnapshot(ctx, chainID, fromHeight, toHeight, true)
+}
+
+func (b *Service) createSnapshot(ctx context.Context, chainID string, fromHeight int64, toHeight int64, async bool) (Meta, error) {
+	var err error
+	if len(chainID) == 0 {
+		return Meta{}, fmt.Errorf("chain id is required")
+	}
+
+	b.log.Infof("creating snapshot for chain %s, from height %d, to height %d", chainID, fromHeight, toHeight)
+
+	conn, err := pgxpool.Connect(ctx, b.connConfig.GetConnectionString())
+	if err != nil {
+		return Meta{}, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	defer conn.Close()
+
+	dbMetaData, err := NewDatabaseMetaData(ctx, b.connConfig)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to get data dump metadata: %w", err)
+	}
+
+	currentSnapshot := NewCurrentSnapshot(chainID, toHeight)
+	historySnapshot := NewHistorySnapshot(chainID, fromHeight, toHeight)
+
+	meta := Meta{
+		ChainID:                  chainID,
+		HeightFrom:               fromHeight,
+		HeightTo:                 toHeight,
+		CurrentStateSnapshotFile: filepath.Join(b.snapshotsPath, currentSnapshot.CompressedFileName()),
+		HistorySnapshotFile:      filepath.Join(b.snapshotsPath, historySnapshot.CompressedFileName()),
+		WriteLockFile:            filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight)),
+		DatabaseVersion:          dbMetaData.DatabaseVersion,
+	}
+
+	currentDataDir, err := b.createUncompressedDataDirectory(currentSnapshot.UncompressedDataDir())
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create uncompressed data directory for current snapshot:%w", err)
+	}
+
+	historyDataDir, err := b.createUncompressedDataDirectory(historySnapshot.UncompressedDataDir())
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create uncompressed data directory for history snapshot:%w", err)
+	}
+
+	copyDataTx, err := conn.Begin(ctx)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to begin copy table data transaction: %w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TIME ZONE 'UTC'")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set timezone to UTC:%w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set transaction isolation level to serilizable:%w", err)
+	}
+
+	writeLockFileName := filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight))
+	_, err = os.Create(writeLockFileName)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create write lock file:%w", err)
+	}
+
+	if async {
+		go func() {
+			err = b.snapshotData(ctx, copyDataTx, currentSnapshot, dbMetaData, historySnapshot, currentDataDir,
```

Comment:
I think we need to do more here than just kick off this all in a go-routine. If someone sigterms the data-node process while a snapshot is still in progress we need to wait for the snapshot to finish before stopping the process, otherwise we'll have a half-baked snapshot hanging around.